{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2024-05-20T01:35:49.852575Z",
     "start_time": "2024-05-20T01:35:49.814457Z"
    }
   },
   "source": "import pandas as pd",
   "outputs": [],
   "execution_count": 28
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-20T01:35:50.906571Z",
     "start_time": "2024-05-20T01:35:50.747054Z"
    }
   },
   "cell_type": "code",
   "source": "!pwd",
   "id": "2d3ef321de9ba2ed",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/ivustianiu/PycharmProjects/mlops-zoomcamp/project\r\n"
     ]
    }
   ],
   "execution_count": 29
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-20T01:35:53.427790Z",
     "start_time": "2024-05-20T01:35:53.395556Z"
    }
   },
   "cell_type": "code",
   "source": [
    "df = pd.read_csv('data/clinvar_annotated.tsv', delimiter='\\t', na_values='-')\n",
    "df.shape"
   ],
   "id": "efe9a79d7461024c",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9058, 20)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 30
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-20T01:35:55.768459Z",
     "start_time": "2024-05-20T01:35:55.761485Z"
    }
   },
   "cell_type": "code",
   "source": "df.iloc[0].T",
   "id": "d6af4a90f23c6909",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant_id             chr1_955677_A_C\n",
       "label                               LB\n",
       "ref                                  A\n",
       "alt                                  C\n",
       "gnomad_AF                     0.000422\n",
       "gnomad_AC                           13\n",
       "gnomad_AN                        30820\n",
       "gnomad_Hom                           0\n",
       "revel                             0.27\n",
       "splice_ai                          0.0\n",
       "amino_acid_change                  E/A\n",
       "biotype                 protein_coding\n",
       "canonical                         True\n",
       "consequence           missense_variant\n",
       "impact                        MODERATE\n",
       "gene_name                         AGRN\n",
       "ncbi_id                         375790\n",
       "strand                               1\n",
       "transcript_id                NM_198576\n",
       "transcript_version                   4\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 31
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-20T01:36:42.294257Z",
     "start_time": "2024-05-20T01:36:42.279180Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Metadata columns (must be excluded): variant_id, gene_name, ncbi_id, transcript_id, transcript_version\n",
    "# Target column: label\n",
    "# Atomic categorical columns: ref, alt, consequence, impact\n",
    "# Compound categorical column (must be split into ref_aa and alt_aa): amino_acid_change\n",
    "\n",
    "df.info()"
   ],
   "id": "c2dbdc94ae36af4c",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 9058 entries, 0 to 9057\n",
      "Data columns (total 20 columns):\n",
      " #   Column              Non-Null Count  Dtype  \n",
      "---  ------              --------------  -----  \n",
      " 0   variant_id          9058 non-null   object \n",
      " 1   label               9058 non-null   object \n",
      " 2   ref                 9058 non-null   object \n",
      " 3   alt                 9058 non-null   object \n",
      " 4   gnomad_AF           9058 non-null   float64\n",
      " 5   gnomad_AC           9058 non-null   int64  \n",
      " 6   gnomad_AN           9058 non-null   int64  \n",
      " 7   gnomad_Hom          9058 non-null   int64  \n",
      " 8   revel               9058 non-null   float64\n",
      " 9   splice_ai           9058 non-null   float64\n",
      " 10  amino_acid_change   8924 non-null   object \n",
      " 11  biotype             9058 non-null   object \n",
      " 12  canonical           9058 non-null   bool   \n",
      " 13  consequence         9058 non-null   object \n",
      " 14  impact              9058 non-null   object \n",
      " 15  gene_name           9058 non-null   object \n",
      " 16  ncbi_id             9058 non-null   int64  \n",
      " 17  strand              9058 non-null   int64  \n",
      " 18  transcript_id       9058 non-null   object \n",
      " 19  transcript_version  9058 non-null   int64  \n",
      "dtypes: bool(1), float64(3), int64(6), object(10)\n",
      "memory usage: 1.3+ MB\n"
     ]
    }
   ],
   "execution_count": 32
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-19T01:36:55.926265Z",
     "start_time": "2024-05-19T01:36:55.918898Z"
    }
   },
   "cell_type": "code",
   "source": "df.loc[df.REVEL_score.isnull(), 'variant_id'].head(10)",
   "id": "91c8345adfecf61b",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     chr1_955677_A_C\n",
       "1     chr1_977396_G_A\n",
       "2     chr1_978801_G_A\n",
       "3     chr1_978820_C_T\n",
       "4     chr1_983695_G_A\n",
       "5     chr1_983713_C_T\n",
       "6     chr1_984280_G_A\n",
       "7     chr1_985125_A_G\n",
       "8    chr1_1149501_C_G\n",
       "9    chr1_1167767_G_A\n",
       "Name: variant_id, dtype: object"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 27
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                                                              17\n",
       "variant_id                                                      chr1_1273745_C_T\n",
       "label                                                                         LB\n",
       "Allele                                                                         T\n",
       "Consequence                                                  5_prime_UTR_variant\n",
       "IMPACT                                                                  MODIFIER\n",
       "SYMBOL                                                                    UBE2J2\n",
       "Gene                                                             ENSG00000160087\n",
       "Feature_type                                                          Transcript\n",
       "Feature                                                       ENST00000349431.11\n",
       "BIOTYPE                                                           protein_coding\n",
       "EXON                                                                         1/7\n",
       "INTRON                                                                       NaN\n",
       "HGVSc                                                                        NaN\n",
       "HGVSp                                                                        NaN\n",
       "cDNA_position                                                              110.0\n",
       "CDS_position                                                                 NaN\n",
       "Protein_position                                                             NaN\n",
       "Amino_acids                                                                  NaN\n",
       "Codons                                                                       NaN\n",
       "Existing_variation                                                           NaN\n",
       "REF_ALLELE                                                                     C\n",
       "UPLOADED_ALLELE                                                              C/T\n",
       "DISTANCE                                                                     NaN\n",
       "STRAND                                                                      -1.0\n",
       "FLAGS                                                                        NaN\n",
       "SYMBOL_SOURCE                                                               HGNC\n",
       "HGNC_ID                                                               HGNC:19268\n",
       "CANONICAL                                                                    YES\n",
       "MANE_SELECT                                                          NM_058167.3\n",
       "MANE_PLUS_CLINICAL                                                           NaN\n",
       "TSL                                                                          1.0\n",
       "APPRIS                                                                        P1\n",
       "MOTIF_NAME                                                                   NaN\n",
       "MOTIF_POS                                                                    NaN\n",
       "HIGH_INF_POS                                                                 NaN\n",
       "MOTIF_SCORE_CHANGE                                                           NaN\n",
       "TRANSCRIPTION_FACTORS                                                        NaN\n",
       "5UTR_annotation                                                              NaN\n",
       "5UTR_consequence                                                             NaN\n",
       "Existing_InFrame_oORFs                                                       0.0\n",
       "Existing_OutOfFrame_oORFs                                                    0.0\n",
       "Existing_uORFs                                                               0.0\n",
       "GO                             GO:0000151:ubiquitin_ligase_complex&GO:0000166...\n",
       "SpliceAI_pred_DP_AG                                                          NaN\n",
       "SpliceAI_pred_DP_AL                                                          NaN\n",
       "SpliceAI_pred_DP_DG                                                          NaN\n",
       "SpliceAI_pred_DP_DL                                                          NaN\n",
       "SpliceAI_pred_DS_AG                                                          NaN\n",
       "SpliceAI_pred_DS_AL                                                          NaN\n",
       "SpliceAI_pred_DS_DG                                                          NaN\n",
       "SpliceAI_pred_DS_DL                                                          NaN\n",
       "SpliceAI_pred_SYMBOL                                                         NaN\n",
       "PHENOTYPES                                                                   NaN\n",
       "REVEL_score                                                                  NaN\n",
       "gnomAD_genomes_POPMAX_AC                                                     NaN\n",
       "gnomAD_genomes_POPMAX_AF                                                     NaN\n",
       "gnomAD_genomes_POPMAX_AN                                                     NaN\n",
       "gnomAD_genomes_POPMAX_nhomalt                                                NaN\n",
       "NMD                                                                          NaN"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>17</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>variant_id</th>\n",
       "      <td>chr1_1273745_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>label</th>\n",
       "      <td>LB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele</th>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Consequence</th>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMPACT</th>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYMBOL</th>\n",
       "      <td>UBE2J2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <td>ENSG00000160087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature_type</th>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature</th>\n",
       "      <td>ENST00000349431.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIOTYPE</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EXON</th>\n",
       "      <td>1/7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INTRON</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGVSc</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGVSp</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cDNA_position</th>\n",
       "      <td>110.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDS_position</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein_position</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Amino_acids</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Existing_variation</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_ALLELE</th>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UPLOADED_ALLELE</th>\n",
       "      <td>C/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DISTANCE</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRAND</th>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLAGS</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYMBOL_SOURCE</th>\n",
       "      <td>HGNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGNC_ID</th>\n",
       "      <td>HGNC:19268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CANONICAL</th>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MANE_SELECT</th>\n",
       "      <td>NM_058167.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MANE_PLUS_CLINICAL</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSL</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APPRIS</th>\n",
       "      <td>P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOTIF_NAME</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIGH_INF_POS</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRANSCRIPTION_FACTORS</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5UTR_annotation</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5UTR_consequence</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Existing_InFrame_oORFs</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Existing_OutOfFrame_oORFs</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Existing_uORFs</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO</th>\n",
       "      <td>GO:0000151:ubiquitin_ligase_complex&amp;GO:0000166...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SpliceAI_pred_DP_AG</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SpliceAI_pred_DP_AL</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SpliceAI_pred_DP_DG</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SpliceAI_pred_DP_DL</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SpliceAI_pred_DS_AG</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SpliceAI_pred_DS_AL</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SpliceAI_pred_DS_DG</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SpliceAI_pred_DS_DL</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SpliceAI_pred_SYMBOL</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHENOTYPES</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REVEL_score</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gnomAD_genomes_POPMAX_AC</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gnomAD_genomes_POPMAX_AF</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gnomAD_genomes_POPMAX_AN</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gnomAD_genomes_POPMAX_nhomalt</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NMD</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 19,
   "source": [
    "df[~df['Consequence'].isin(\n",
    "    ['intron_variant', 'upstream_gene_variant', 'downsream_gene_variant', 'regulatory_region_variant']\n",
    ")].head(1).T"
   ],
   "id": "c70245d3ec7ab48d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "ccee8c38174cebf5"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
